tgui/packages/tgui/interfaces/NtosMessenger.tsx

```diff
@@ -0,0 +1,469 @@
+          </Box>
+        </Stack>
+      </Section>
+      {!!photo && (
+        <Stack vertical>
+          <Section fill textAlign="center">
+            <Icon name="camera" mr={1} />
+            Current Photo
+          </Section>
+          <Section align="center" mb={1}>
+            <Button onClick={() => act('PDA_clearPhoto')}>
+              <Box mt={1} as="img" src={photo} />
+            </Button>
+          </Section>
+        </Stack>
+      )}
+      <Stack vertical>
+        <Section fill textAlign="center">
+          <Icon name="address-card" mr={1} />
+          Detected Messengers
+          <Input
+            width="220px"
+            placeholder="Search by name or job..."
+            value={searchUser}
+            onInput={(_e: any, value: string) => setSearchUser(value)}
+            mx={1}
+            ml={27}
+          />
+        </Section>
+      </Stack>
+      <Stack vertical fill mt={1}>
+        <Section fill scrollable>
+          <Stack vertical pb={1}>
+            {users.length === 0 && 'No users found'}
+            {users.map((messenger) => (
+              <Button
+                key={messenger.ref}
+                fluid
+                onClick={() => {
+                  act('PDA_viewMessages', { ref: messenger.ref });
+                }}>
+                {messenger.name} ({messenger.job})
+              </Button>
+            ))}
+          </Stack>
+        </Section>
+        {!!can_spam && <SendToAllModal />}
+      </Stack>
+      {noId && <NoIDDimmer />}
+    </Stack>
+  );
+};
+
+type ChatMessageProps = {
+  isSelf: boolean;
+  msg: string;
+  everyone?: boolean;
+  photoPath?: string;
+};
+
+const ChatMessage: SFC<ChatMessageProps> = (props: ChatMessageProps) => {
+  const { msg, everyone, isSelf, photoPath } = props;
+  const text = {
+    __html: sanitizeText(msg),
+  };
+
+  return (
+    <Box className={`NtosMessenger__ChatMessage${isSelf ? '__outgoing' : ''}`}>
+      <Box
+        className="NtosMessenger__ChatMessage__content"
+        dangerouslySetInnerHTML={text}
+      />
+      {photoPath !== null && <Box as="img" src={photoPath} />}
+      {everyone && (
+        <Box className="NtosMessenger__ChatMessage__everyone">
+          Sent to everyone
+        </Box>
+      )}
+    </Box>
+  );
+};
+
+type ChatScreenProps = {
+  onReturn: () => void;
+  recp: NtMessenger;
+  msgs: NtMessage[];
+};
+
+type ChatScreenState = {
+  msg: string;
+  canSend: boolean;
+};
+
+class ChatScreen extends Component<ChatScreenProps, ChatScreenState> {
+  scrollRef: RefObject<HTMLDivElement>;
+  state: ChatScreenState = {
+    msg: '',
+    canSend: true,
+  };
+
+  constructor(props: ChatScreenProps) {
+    super(props);
+
+    this.scrollRef = createRef();
+
+    this.scrollToBottom = this.scrollToBottom.bind(this);
+    this.handleMessageInput = this.handleMessageInput.bind(this);
+    this.handleSendMessage = this.handleSendMessage.bind(this);
+  }
+
+  componentDidMount(): void {
```

Comment:
i don't hate that you've typed the returns but this isn't needed. unless there's a particularly good reason, i prefer that return types are inferred